pkg/cri/resource-manager/resource-manager.go

```diff
@@ -53,188 +53,290 @@ type resmgr struct {
-	}
-	if m.relay, err = relay.NewRelay(ropts); err != nil {
-		return nil, resmgrError("failed to create resource manager: %v", err)
-	}
-
-	copts := cache.Options{
-		CacheDir: opt.RelayDir,
-	}
-	if m.cache, err = cache.NewCache(copts); err != nil {
-		return nil, resmgrError("failed to create resource manager: %v", err)
-	}
-
-	if err = m.loadInitialConfig(); err != nil {
-		return nil, resmgrError("failed to load initial configuration: %v", err)
-	}
-
-	if policy.ActivePolicy() != m.cache.GetActivePolicy() {
-		if m.cache.GetActivePolicy() != "" {
-			return nil, resmgrError("trying to load cache with policy %s for active policy %s",
-				m.cache.GetActivePolicy(), policy.ActivePolicy())
-		}
-		m.cache.SetActivePolicy(policy.ActivePolicy())
+	if err := m.loadConfig(); err != nil {
+		return nil, err
 	}
 
-	policyOpts := &policy.Options{
-		AgentCli: m.agent,
-	}
-	if m.policy, err = policy.NewPolicy(m.cache, policyOpts); err != nil {
-		return nil, resmgrError("failed to create resource manager: %v", err)
+	if err := m.setupConfigServer(); err != nil {
+		return nil, err
 	}
 
-	if err = m.relay.Setup(); err != nil {
-		return nil, resmgrError("failed to create resource manager: %v", err)
+	if err := m.setupPolicy(); err != nil {
+		return nil, err
 	}
 
-	if err = m.setupRequestProcessing(); err != nil {
-		return nil, resmgrError("failed to create resource manager: %v", err)
+	if err := m.setupRelay(); err != nil {
+		return nil, err
 	}
 
-	if m.control, err = control.NewControl(); err != nil {
-		return nil, resmgrError("failed to create controller: %v", err)
+	if err := m.setupRequestProcessing(); err != nil {
+		return nil, err
 	}
 
-	if opt.ForceConfig != "" {
-		m.Warn("using forced configuration %s, not starting config server...", opt.ForceConfig)
-	} else {
-		if m.configServer, err = config.NewConfigServer(m.SetConfig); err != nil {
-			return nil, resmgrError("failed to create resource manager: %v", err)
-		}
+	if err := m.setupControllers(); err != nil {
+		return nil, err
 	}
 
 	return m, nil
 }
 
 // Start starts the resource manager.
 func (m *resmgr) Start() error {
-	m.Info("starting resource manager...")
+	m.Info("starting...")
 
 	m.Lock()
 	defer m.Unlock()
 
-	if opt.ForceConfig == "" {
-		if err := m.configServer.Start(opt.ConfigSocket); err != nil {
-			return resmgrError("failed to start config-server: %v", err)
-		}
-	}
-
-	if err := m.activateRequestProcessing(); err != nil {
+	if err := m.startControllers(); err != nil {
 		return err
 	}
 
-	if err := m.control.StartStopControllers(m.cache, m.relay.Client()); err != nil {
-		return resmgrError("failed to start controller: %v", err)
+	if err := m.startRequestProcessing(); err != nil {
+		return err
 	}
 
 	if err := m.relay.Start(); err != nil {
-		return resmgrError("failed to start relay: %v", err)
+		return resmgrError("failed to start CRI relay: %v", err)
 	}
 
-	if m.conf != nil {
-		m.cache.SetConfig(m.conf)
+	if opt.ForceConfig == "" {
```

Comment:
question: this is already checked in `setupConfigServer()`. Should/could we check `m.configServer != nil` here, instead? To me this would prolly be a bit more 
 readable/understandable but I guess this is very subtle and a personal preference.

Reply:
> Should/could we check `m.configServer != nil` here, instead?

Well, this is an interface, so I think that wouldn't work (I mean testing nilness without reflect). If we want to do that, we need to change it to a *config.Server. Then we could check for nil, but then we'd always need to (*m.configServer).Func() it, IIRC.